tmail-backend/blob/secondary-blob-store/src/main/java/com/linagora/tmail/blob/secondaryblobstore/SecondaryBlobStoreDAO.java

```diff
@@ -0,0 +1,176 @@
+    }
+
+    @Override
+    public Mono<Void> delete(BucketName bucketName, BlobId blobId) {
+        return Flux.merge(asSavingStatus(firstBlobStoreDAO.delete(bucketName, blobId)),
+                asSavingStatus(secondBlobStoreDAO.delete(bucketName, blobId)))
+            .collectList()
+            .handle(this::merge);
+    }
+
+    @Override
+    public Mono<Void> delete(BucketName bucketName, Collection<BlobId> blobIds) {
+        return Flux.merge(asSavingStatus(firstBlobStoreDAO.delete(bucketName, blobIds)),
+                asSavingStatus(secondBlobStoreDAO.delete(bucketName, blobIds)))
+            .collectList()
+            .handle(this::merge);
+    }
+
+    @Override
+    public Mono<Void> deleteBucket(BucketName bucketName) {
+        return Flux.merge(asSavingStatus(firstBlobStoreDAO.deleteBucket(bucketName)),
+                asSavingStatus(secondBlobStoreDAO.deleteBucket(bucketName)))
+            .collectList()
+            .handle(this::merge);
+    }
+
+    @Override
+    public Flux<BucketName> listBuckets() {
+        return Flux.from(firstBlobStoreDAO.listBuckets())
+            .onErrorResume(ex -> {
+                LOGGER.warn("Fail to list buckets from the first blob store. Use second blob store", ex);
+                return Flux.from(secondBlobStoreDAO.listBuckets());
+            });
+    }
+
+    @Override
+    public Flux<BlobId> listBlobs(BucketName bucketName) {
+        return Flux.from(firstBlobStoreDAO.listBlobs(bucketName))
+            .onErrorResume(ex -> {
+                LOGGER.warn("Fail to list blobs from the first blob store with bucket name {}. Use second blob store", bucketName.asString(), ex);
+                return Flux.from(secondBlobStoreDAO.listBlobs(bucketName));
+            });
+    }
+
+    private Mono<SavingStatus> asSavingStatus(Publisher<Void> publisher) {
+        return Mono.from(publisher).then(Mono.just(SavingStatus.success()))
+            .onErrorResume(e -> Mono.just(SavingStatus.failure(e)));
+    }
+
+    private void merge(List<SavingStatus> savingStatuses, SynchronousSink<Void> sink) {
+        Preconditions.checkArgument(savingStatuses.size() == 2);
+        boolean bothSucceeded = savingStatuses.get(0).isSuccess() && savingStatuses.get(1).isSuccess();
+        boolean bothFailed = !savingStatuses.get(0).isSuccess() && !savingStatuses.get(1).isSuccess();
+        if (bothSucceeded) {
+            sink.complete();
+            return;
+        }
+        if (bothFailed) {
+            sink.error(new ObjectStoreException("Failure to save in both blobStore. First exception was:", savingStatuses.getFirst().e.get()));
+            return;
+        }
+        if (savingStatuses.get(0).isSuccess()) {
+            LOGGER.warn("Failure to save in secondary blobStore", savingStatuses.get(1).e.get());
+        } else {
+            LOGGER.warn("Failure to save in primary blobStore", savingStatuses.get(0).e.get());
+        }
+        sink.complete();
```

Comment:
More pretty
```suggestion
        List<SavingStatus> successList = savingStatuses.stream()
            .filter(SavingStatus::isSuccess)
            .toList();

        switch (successList.size()) {
            case 0 -> sink.error(new ObjectStoreException("Failure to save in both blobStore", savingStatuses.getFirst().e().get()));
            case 2 -> sink.complete();
            case 1 -> {
                LOGGER.warn("Failure to save in one blobStore", savingStatuses.getFirst().e().get());
                sink.complete();
            }
        }
```

Reply:
the old code let us know which blobstore fail. so keep the old code

Reply:
Yep this is a bit boiler plate but this was intended.